### PR TITLE
agent: Add support to online vCPUs

### LINKED ISSUE #189: Support hot plugged vCPUs
Online vCPUs when they are hot plugged

### PR DESCRIPTION
**Pull request by devimc**
this patch is to online vCPU in the POD when a new vCPU is
hot added by the runtime

fixes #189

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

File: agent.go [hunk 1]
```diff
@@ -72,6 +72,7 @@ const (
 	cannotGetTimeMsg   = "Failed to get time for event %s:%v"
 	pciBusRescanFile   = "/sys/bus/pci/rescan"
 	pciBusMode         = 0220
+	cpuMode            = 0644
```
Comment (sboeuf) [inline agent.go]: I think `0600` should be enough.

File: agent.go [hunk 2]
```diff
@@ -576,6 +577,11 @@ func (p *pod) listenToUdevEvents(done chan struct{}) {
 				continue
 			}
 
+			if d.Subsystem() == "cpu" {
```
Comment (sboeuf) [inline agent.go]: Might worth putting `"block"` `"cpu"` and `"add"` into some constants.

File: agent.go (hunk 1 above)
Comment (jodh-intel) [reply agent.go]: Yeah, maybe even `0200`?

File: agent.go [hunk 3]
```diff
@@ -593,6 +599,16 @@ func (p *pod) listenToUdevEvents(done chan struct{}) {
 	}()
 }
 
+func (p *pod) onlineVCPU(syspath string) {
+	onlinePath := filepath.Join(syspath, "online")
```
Comment (sboeuf) [inline agent.go:566]: I assume the syspath you're gonna get here is going to be something like:
```
/sys/devices/system/cpu/cpuN
```
Right ?
Could be nice to check the path is the expected pattern before to concatenate it with `"online"`.

File: agent.go (hunk 1 above)
Comment (devimc) [reply agent.go]: this file is used to set and get the cpu status, ```0600``` should be enough

File: agent.go [hunk 4]
```diff
@@ -593,6 +599,16 @@ func (p *pod) listenToUdevEvents(done chan struct{}) {
...
+	if err := ioutil.WriteFile(onlinePath, []byte("1"), cpuMode); err != nil {
+		fieldLogger := agentLog.WithField("subsystem", "udevlistener")
+		fieldLogger.WithFields(logrus.Fields{
```
Comment (sboeuf) [inline agent.go]: You can simplify this with something like that:
```go
agentLog.WithFields(logrus.Fields{
        "subsystem": "udevlistener",
        "subsystem": "cpu",
}).Errorf("failed to online vCPU %s: %s", syspath, err)
```

**Review by sboeuf — COMMENTED** [review]
A few comments but the change looks good.

File: agent.go (hunk 3 above)
Comment (devimc) [reply agent.go:566]: I'd prefer to just check if the file exist

**Comment by devimc** [comment]
@sboeuf  @jodh-intel  changes applied, thanks

File: agent.go (hunk 3 above)
Comment (sboeuf) [reply agent.go:566]: Sounds good.

**Review by sboeuf — APPROVED** [review]

**Comment by sboeuf** [comment]
LGTM

**Comment by jodh-intel** [comment]
lgtm

[Approved with PullApprove](https://pullapprove.com/clearcontainers/agent/pull-request/190/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/agent)

**Comment by jodh-intel** [comment]
Restarted 17.04 CI - looks like a network timeout talking to the Ubuntu servers.

File: agent.go [hunk 5]
```diff
@@ -531,7 +536,12 @@ func (p *pod) listenToUdevEvents(done chan struct{}) {
 
 			// Ignore udev events for block devices, these are handled by another go-routine in waitForBlockDevice
 			// that mounts block devices at the expected location provided on the command line.
-			if d.Subsystem() == "block" {
+			if d.Subsystem() == blockSubsystem {
+				continue
+			}
+
+			if d.Subsystem() == cpuSubsystem {
+				p.onlineVCPU(d.Syspath())
```
Comment (amshinde) [inline agent.go:544]: if this operation takes some time, you can consider doing it in a separate go-routine.

Comment (devimc) [reply agent.go:544]: this function only writes ```1``` in ```/sys/devices/system/cpu/cpuX/online```

**Pull request merged**